contracts/src/v0.8/dev/transmission/4337/Paymaster.sol

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+import "../../vendor/entrypoint/interfaces/IPaymaster.sol";
+import "./SCALibrary.sol";
+import "../../vendor/entrypoint/core/Helpers.sol";
+import "../../../interfaces/LinkTokenInterface.sol";
+import "../../../interfaces/AggregatorV3Interface.sol";
+import "./SCALibrary.sol";
+
+/// @dev LINK token paymaster implementation.
+/// TODO: more documentation.
+contract Paymaster is IPaymaster {
+  error OnlyCallableFromLink();
+  error InvalidCalldata();
+  error UserOperationAlreadyTried(bytes32 userOpHash);
+  error InsufficientFunds(uint256 juelsNeeded, uint256 subscriptionBalance);
+
+  LinkTokenInterface public immutable LINK;
+  AggregatorV3Interface public immutable LINK_ETH_FEED;
+
+  struct Config {
+    uint32 stalenessSeconds;
+    int256 fallbackWeiPerUnitLink;
+  }
+  Config public s_config;
+
+  mapping(bytes32 => bool) userOpHashMapping;
+  mapping(address => uint256) subscriptions;
+
+  constructor(LinkTokenInterface linkToken, AggregatorV3Interface linkEthFeed) {
+    LINK = linkToken;
+    LINK_ETH_FEED = linkEthFeed;
+  }
+
+  function setConfig(uint32 stalenessSeconds, int256 fallbackWeiPerUnitLink) external {
+    s_config = Config({stalenessSeconds: stalenessSeconds, fallbackWeiPerUnitLink: fallbackWeiPerUnitLink});
+  }
+
+  function onTokenTransfer(address /* _sender */, uint256 _amount, bytes calldata _data) external {
+    if (msg.sender != address(LINK)) {
+      revert OnlyCallableFromLink();
+    }
+    if (_data.length != 32) {
+      revert InvalidCalldata();
+    }
+
+    address subscription = abi.decode(_data, (address));
+    subscriptions[subscription] += _amount;
+  }
+
+  function validatePaymasterUserOp(
+    UserOperation calldata userOp,
+    bytes32 userOpHash,
+    uint256 maxCost
+  ) external returns (bytes memory context, uint256 validationData) {
+    if (userOpHashMapping[userOpHash]) {
+      revert UserOperationAlreadyTried(userOpHash);
+    }
+
+    uint256 extraCostJuels = handleExtraCostJuels(userOp);
+    uint256 costJuels = getCostJuels(maxCost) + extraCostJuels;
+    if (subscriptions[userOp.sender] < costJuels) {
+      revert InsufficientFunds(costJuels, subscriptions[userOp.sender]);
+    }
+
+    userOpHashMapping[userOpHash] = true;
+    return (abi.encode(userOp.sender, extraCostJuels), _packValidationData(false, 0, 0)); // success
+  }
+
+  /// @dev Calculates any extra LINK cost for the user operation, based on the funding type passed to the
+  /// @dev paymaster. Handles funding the LINK token funding described in the user operation.
+  /// TODO: add logic for subscription top-up.
+  function handleExtraCostJuels(UserOperation calldata userOp) internal returns (uint256 extraCost) {
+    if (userOp.paymasterAndData.length == 20) {
```

Comment:
does the check need to be `<=20` in case user sends unexpected data?

Reply:
https://github.com/smartcontractkit/chainlink/blob/transmission-don-contracts/contracts/src/v0.8/dev/vendor/entrypoint/core/EntryPoint.sol#L342-L346
The EntryPoint ensures length is at least 20 here: https://github.com/smartcontractkit/chainlink/blob/transmission-don-contracts/contracts/src/v0.8/dev/vendor/entrypoint/core/EntryPoint.sol#L342-L346